File: samples/SampleIssueTracker/JsonContext.cs
using System.Text.Json.Serialization;
using SampleIssueTracker.Issues.Domain;

namespace SampleIssueTracker;

/// <summary>
/// Metano transpiles this class into a <c>JsonContext extends SerializerContext</c>
/// on the TypeScript side, with a lazy <see cref="Metano.Runtime.TypeSpec{T}"/>
/// getter per <c>[JsonSerializable]</c> type.
///
/// <para>
/// Exercises the full descriptor surface that Phase 5 validation cares about:
/// <list type="bullet">
///   <item>branded primitives (<see cref="IssueId"/>, <see cref="UserId"/>)</item>
///   <item>string enums (<see cref="IssuePriority"/>, <see cref="IssueStatus"/>, <see cref="IssueType"/>)</item>
///   <item>nullable fields (<c>UserId? AssigneeId</c>, <c>string? SprintKey</c>)</item>
///   <item>Temporal types (<see cref="System.DateTimeOffset"/>)</item>
///   <item>decimal (for the custom-converter round-trip test)</item>
///   <item>array of nested refs (<see cref="Comment"/>[])</item>
/// </list>
/// </para>
/// </summary>
[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase)]
[JsonSerializable(typeof(IssueSnapshot))]
[JsonSerializable(typeof(Comment))]
public partial class JsonContext : JsonSerializerContext;

File: targets/js/sample-issue-tracker/src/json-context.ts
import { Temporal } from "@js-temporal/polyfill";
import { SerializerContext, type TypeSpec } from "metano-runtime/system/json";
import { Decimal } from "decimal.js";
import { Comment, IssueId, IssuePriority, IssueSnapshot, IssueStatus, IssueType } from "#/issues/domain";
import { UserId } from "#/shared-kernel";

export class JsonContext extends SerializerContext {
  private static readonly _default: JsonContext = new JsonContext();

  private _issueSnapshot?: TypeSpec<IssueSnapshot>;

  private _comment?: TypeSpec<Comment>;

  static get default(): JsonContext {
    return this._default;
  }

  get issueSnapshot(): TypeSpec<IssueSnapshot> {
    return this._issueSnapshot ??= this.createSpec({
      type: IssueSnapshot,
      factory: (p: Record<string, unknown>) => new IssueSnapshot(p.id as IssueId, p.title as string, p.description as string, p.type as IssueType, p.priority as IssuePriority, p.status as IssueStatus, p.assigneeId as UserId | null, p.sprintKey as string | null, p.createdAt as Temporal.ZonedDateTime, p.updatedAt as Temporal.ZonedDateTime, p.estimatedHours as Decimal, p.comments as Comment[]),
      properties: [
        {
          ts: "id",
          json: "id",
          type: { kind: "branded", create: IssueId.create },
        },
        {
          ts: "title",
          json: "title",
          type: { kind: "primitive" },
        },
        {
          ts: "description",
          json: "description",
          type: { kind: "primitive" },
        },
        {
          ts: "type",
          json: "type",
          type: { kind: "enum", values: IssueType },
        },
        {
          ts: "priority",
          json: "priority",
          type: { kind: "enum", values: IssuePriority },
        },
        {
          ts: "status",
          json: "status",
          type: { kind: "enum", values: IssueStatus },
        },
        {
          ts: "assigneeId",
          json: "assigneeId",
          type: {
            kind: "nullable",
            inner: { kind: "branded", create: UserId.create },
          },
          optional: true,
        },
        {
          ts: "sprintKey",
          json: "sprintKey",
          type: {
            kind: "nullable",
            inner: { kind: "primitive" },
          },
          optional: true,
        },
        {
          ts: "createdAt",
          json: "createdAt",
          type: { kind: "temporal", parse: Temporal.ZonedDateTime.from },
        },
        {
          ts: "updatedAt",
          json: "updatedAt",
          type: { kind: "temporal", parse: Temporal.ZonedDateTime.from },
        },
        {
          ts: "estimatedHours",
          json: "estimatedHours",
          type: { kind: "decimal" },
        },
        {
          ts: "comments",
          json: "comments",
          type: {
            kind: "array",
            element: {
              kind: "ref",
              spec: () => this.comment,
            },
          },
        },
      ],
    });
  }

  get comment(): TypeSpec<Comment> {
    return this._comment ??= this.createSpec({
      type: Comment,
      factory: (p: Record<string, unknown>) => new Comment(p.authorId as UserId, p.message as string, p.createdAt as Temporal.ZonedDateTime, p.isSystem as boolean),
      properties: [
        {
          ts: "authorId",
          json: "authorId",
          type: { kind: "branded", create: UserId.create },
        },
        {
          ts: "message",
          json: "message",
          type: { kind: "primitive" },
        },
        {
          ts: "createdAt",
          json: "createdAt",
          type: { kind: "temporal", parse: Temporal.ZonedDateTime.from },
        },
        {
          ts: "isSystem",
          json: "isSystem",
          type: { kind: "primitive" },
        },
      ],
    });
  }
}
